### DISCUSSION TITLE
Custom webservice problem with XML HL7 message

### DISCUSSION THREAD

**Discussion by dbarreno** [nextgenhealthcare/connect#5701] [Q&A] [answered]
Hello to the whole Mirth Community,

I have created a custom webservice.
This webservice works properly only when the HL7-XML message I need to extract to send Mirth is included into a CDATA tag.
If the XML is not included in a CDATA tag the message is empty.

This is what I mean.

If the message is like this, it works:
`<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:sas="http://SAS.Servicios/">
   <soapenv:Header/>
   <soapenv:Body>
      <sas:S014>
         
         <MensEntrada>
                <![CDATA[<SIU_S12 xmlns="urn:hl7-org:v2xml">
                    <MSH>
                          <MSH.1>|</MSH.1>
                     ................
                </SIU_S12.RESOURCES>
                </SIU_S12>]]>
</MensEntrada>
      </sas:S014>
   </soapenv:Body>
</soapenv:Envelope>`
![image](https://user-images.githubusercontent.com/4416342/221808463-7c99b8c6-8152-4b9b-8491-1ea5d5b94998.png)

If the message is like this, the result is a blank message. 

`<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:sas="http://SAS.Servicios/">
   <soapenv:Header/>
   <soapenv:Body>
      <sas:S014>
         
         <MensEntrada>
                <SIU_S12 xmlns="urn:hl7-org:v2xml">
                    <MSH> 
                    <MSH.1>|</MSH.1>
                     ................
                </SIU_S12.RESOURCES>
                </SIU_S12>
</MensEntrada>
      </sas:S014>
   </soapenv:Body>
</soapenv:Envelope>`

![image](https://user-images.githubusercontent.com/4416342/221808614-4239367a-688f-4590-bf49-7e935fa49f94.png)

This is the JAVA code I am using with the custom webservice:

`@WebService(		targetNamespace = "http://SAS.Servicios/",
			serviceName = "S014Service",
			name = "S014Service", 
			portName = "S014ServiceSoap")
public class S014Service extends AcceptMessage {

	public S014Service(WebServiceReceiver webServiceMessageReceiver) {
		super(webServiceMessageReceiver);
	}
	@WebMethod(operationName = "S014")
        @WebResult(name = "S014Response")
	public String S014(@WebParam(name="MensEntrada") String message) {
		return webServiceReceiver.processData(message);
	}
}`
The problem is the customer is sending the messages without the CDATA tag, and can not change anything on his side, so I have to solve this issue but I have no idea what I am doing wrong.

**Comment by ab-20-m** [comment]
I'm not experienced in JAX-WS and JAXB, but the results you are experiencing would seem to be the default deserialization behavior. I think to get what you want; you will need to create a custom class annotated with the appropriate JAXB annotations. Then make that custom class the input parameter to the service method.
So modified service method:
```
@WebMethod(operationName = "S014")
@WebResult(name = "S014Response")
public String S014(MensEntradaWrapper wrapper) 
{
	String contents = wrapper.getContentsAsString();
	return webServiceReceiver.processData(contents);
}
````

For the custom class/wrapper:
```
@XmlRootElement(name = "MensEntrada")
public class MensEntradaWrapper
{
	//Assumes only 1 child element for content
	@XmlAnyElement
	private Element content;

	public String getContentAsString() 
	{
		StringBuilder builder = new StringBuilder();
		StringWriter writer = new StringWriter();
	        Transformer transformer = TransformerFactory.newInstance().newTransformer();
        	transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
	        transformer.transform(new DOMSource(contents), new StreamResult(writer));
        	builder.append(writer.toString());
		return builder.toString();
	}
}
```

Reference Links:
[StackOverflow JAX-WS web service - Get portion of input XML in String to a field](https://stackoverflow.com/questions/19935528/jax-ws-web-service-get-portion-of-input-xml-in-string-to-a-field)
[JAXB @XmlAnyElement example](http://lstierneyltd.com/blog/development/tips/jaxb-xmlanyelement-example/)

**Comment by dbarreno** [comment]
Hi @ab-20-m ,

Thank you very much for the information. I have tested and make some modifications on this way.
The code is compiled in Netbeans properly, but have this error when Mirth try to start the channel:

![image](https://user-images.githubusercontent.com/4416342/222378108-915d5ee9-bb78-4241-8cad-3dde6a8d7f0f.png)

Error is :
javax.xml.transform.SourceLocator is an interface and JAXB can not handle interfaces.

Do you have any idea how to solve it?

**Reply by ab-20-m** [reply]
Unfortunately, I don't have slightest idea how to resolve that error. If you have the WSDL the sending system is using, you should be able to generate the code for the Java service side and compare to what you have now. It might be a simple missing annotation or an attribute on an annotation is required.

**Reply by ab-20-m** [reply]
I was able to look a little more into this. If I allowed the XML Transform exceptions to bubble up and get thrown by the service method I see the same error. Looks like JAXB and JAX-WS can't map those exceptions to Faults.
Below code should get you further along. Adapt as needed.

Service Definition 2.0
```
@WebMethod(operationName = "S014")
@WebResult(name = "S014Response")
public String S014(@WebParam(name="MensEntrada") MensEntradaWrapper wrapper) 
{
	String contents = wrapper.getContentsAsString();
	return webServiceReceiver.processData(contents);
}
```

Wrapper/DTO 2.0
```
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "MensEntrada", propOrder = { "any" })
public class MensEntradaWrapper
{
	//Assumes only 1 child element for content
	@XmlAnyElement(lax = true)
	private Object any;

	public String getContentAsString() 
	{
		StringBuilder builder = new StringBuilder();
		try{
			StringWriter writer = new StringWriter();
		        Transformer transformer = TransformerFactory.newInstance().newTransformer();
	        	transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
		        transformer.transform(new DOMSource((Node)any), new StreamResult(writer));
	        	builder.append(writer.toString());
		}catch(TransformerConfigurationException | TransformerFactoryConfigurationError e) {
			return "Invalid XML";
		} catch(TransformerException e) {
			return "Invalid XML";
		}
		return builder.toString();
	}
}
```

**Comment by dbarreno** [comment, marked as answer]
Hi,
Thank you for the answer.
Now Mirth can create the channel, but there is other error when receive the message:
This is the webservice response:

  <faultcode>S:Server</faultcode>
         <faultstring>class com.sun.org.apache.xerces.internal.dom.ElementNSImpl cannot be cast to class javax.xml.soap.Node (com.sun.org.apache.xerces.internal.dom.ElementNSImpl is in module java.xml of loader 'bootstrap'; javax.xml.soap.Node is in unnamed module of loader java.net.URLClassLoader @4ae82894)</faultstring>

**Reply by ab-20-m** [reply]
Check your import statements. The cast in the transformer call should be to `org.w3c.dom.Node` not `javax.xml.soap.Node`.

**Comment by dbarreno** [comment]
Great @ab-20-m ! It works.
Thank you very much!!!
